In [1]:
%reload_ext autoreload
%autoreload 2

# Setup


In [79]:
import pandas as pd

import warnings
%matplotlib inline
warnings.filterwarnings('ignore')
pd.set_option('display.max_colwidth',200)

# Load Dataset

In [80]:
train_data = pd.read_csv("../data/SMM4H_2026/train_data_SMM4H_2026_Task_1.csv")
dev_data = pd.read_csv("../data/SMM4H_2026/dev_data_SMM4H_2026_Task_1.csv")

# Data Exploration

In [81]:
train_data.head()

,id,text,label,origin,type,language,split
0,ru_8872,"знаете эти надписи на баночках с лекарствами звучащие как названия фильмов с подзаголовками? типа «ЦИПРОМЕД ципрофлоксацин 0.3%» вы спросите, что это такое я хз. наверное, киберпанк",0.0,RuDReC,tweet,ru,SMM4H_26_train
1,ru_5231,"Знаю лично людей, которые лечились этим препаратом и успешно избавились от гепатита с. Чтобы лечение было эффективным к этому лекарству еще назначают таблетки рибавирин.",0.0,RuDReC,sentence,ru,SMM4H_26_train
2,zh_1298,我6月19号胃痛送去急诊科注射了两针颅通定针来止痛，结果6月23号就查出怀孕了，可是我月经应该是6月27来的，请问这两针对胎儿是否有影响，能不能生下来，请一定要帮帮我，我真的好着急第一次问题补充:病史:打过一胎第二次问题补充:如果生下来是否对孩子不利，能不能生下来第三次问题补充:药物名称:罗通定〔典〕,0.0,SMM4H_2026,120ask,zh,SMM4H_26_train
3,ja_13288,'ルルド先生が待機終了しました。06月21日22:30',0.0,SMM4H_2026,tweet,ja,SMM4H_26_train
4,en_14557,the following tweets are brought to you by seroquel™,0.0,previous_SMM4H,tweet,en,SMM4H_26_train


In [82]:
dev_data.head()

,id,text,label,origin,type,language,split
0,de_456,"Hallo <user>, DCIS = duktales Carcinoma in situ. Es gibt noch das LCIS = lobuläres Carcinoma in situ. In situ bedeutet an Ort und Stelle. Das sind Brustkrebsvorstufen. Manche Ärzte sprechen lieber...",0.0,LREC2022,forum post,de,SMM4H_26_dev
1,ja_1898,'24時間経過。38.2℃。ナロン錠効果。',0.0,SMM4H_2026,tweet,ja,SMM4H_26_dev
2,ja_11752,'ルルはパパとお家帰ったよ〜🤣今頃お家でご飯食べてそう🍚本日も宜しくお願いします✨💕',0.0,SMM4H_2026,tweet,ja,SMM4H_26_dev
3,ja_14355,'ルル(黒)はヤンチャな乙女ですからね🤣🤣🤣',0.0,SMM4H_2026,tweet,ja,SMM4H_26_dev
4,ru_7420,@pharmazzzevt Акард рабочая штука или фуфломицин? Назначили пить после курса Ксарелто и очистки легких от ТЭЛА.,0.0,RuDReC,tweet,ru,SMM4H_26_dev


In [83]:
train_data.info()

<class 'pandas.DataFrame'>
RangeIndex: 46737 entries, 0 to 46736
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   id        46737 non-null  str    
 1   text      46737 non-null  str    
 2   label     46737 non-null  float64
 3   origin    46737 non-null  str    
 4   type      46737 non-null  str    
 5   language  46737 non-null  str    
 6   split     46737 non-null  str    
dtypes: float64(1), str(6)
memory usage: 12.1 MB


In [84]:
dev_data.info()

<class 'pandas.DataFrame'>
RangeIndex: 8033 entries, 0 to 8032
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   id        8033 non-null   str    
 1   text      8033 non-null   str    
 2   label     8033 non-null   float64
 3   origin    8033 non-null   str    
 4   type      8033 non-null   str    
 5   language  8033 non-null   str    
 6   split     8033 non-null   str    
dtypes: float64(1), str(6)
memory usage: 2.5 MB


In [85]:
train_data["split"].value_counts()

split
SMM4H_26_train    46737
Name: count, dtype: int64

In [86]:
dev_data["split"].value_counts()

split
SMM4H_26_dev    8033
Name: count, dtype: int64

In [87]:
train_data["label"].value_counts()

label
0.0    43745
1.0     2992
Name: count, dtype: int64

In [88]:
dev_data["label"].value_counts()

label
0.0    7525
1.0     508
Name: count, dtype: int64

In [89]:
train_data["language"].value_counts()

language
en    17128
ja    14208
ru    10695
zh     2248
de     1482
fr      976
Name: count, dtype: int64

In [90]:
dev_data["language"].value_counts()

language
ja    3045
ru    2669
en     888
de     634
fr     418
zh     379
Name: count, dtype: int64

In [91]:
train_data["type"].value_counts()

type
tweet         37881
sentence       3754
forum post     2458
120ask         2248
review          396
Name: count, dtype: int64

In [92]:
dev_data["type"].value_counts()

type
tweet         5569
forum post    1052
sentence       934
120ask         379
review          99
Name: count, dtype: int64

# Task 1.1: Language Selection, and Splits

In [93]:
train_langs = ["en", "ru"]   # English and Russian for Training
zero_shot_lang = "de"       # German for Evaluation
selected_langs = train_langs + [zero_shot_lang]

In [94]:
train_val = train_data[train_data["language"].isin(selected_langs)].copy()
train_val

,id,text,label,origin,type,language,split
0,ru_8872,"знаете эти надписи на баночках с лекарствами звучащие как названия фильмов с подзаголовками? типа «ЦИПРОМЕД ципрофлоксацин 0.3%» вы спросите, что это такое я хз. наверное, киберпанк",0.0,RuDReC,tweet,ru,SMM4H_26_train
1,ru_5231,"Знаю лично людей, которые лечились этим препаратом и успешно избавились от гепатита с. Чтобы лечение было эффективным к этому лекарству еще назначают таблетки рибавирин.",0.0,RuDReC,sentence,ru,SMM4H_26_train
4,en_14557,the following tweets are brought to you by seroquel™,0.0,previous_SMM4H,tweet,en,SMM4H_26_train
5,ru_34,"Я если честно боюсь ребенку столько таблеток в день давать, поэтому принимали обычно три раза в день всего пять дней.",0.0,RuDReC,sentence,ru,SMM4H_26_train
6,en_11127,"rt @USER_______: my breakfast consists of vyvanse, coffee, and more vyvanse.",0.0,previous_SMM4H,tweet,en,SMM4H_26_train
...,...,...,...,...,...,...,...
46727,en_8967,@USER_______ how long hsve you been on humira,0.0,previous_SMM4H,tweet,en,SMM4H_26_train
46729,ru_11694,"Возможно, у тебя просто побочки из-за того, что ты только садишься на эти колеса. Это пройдет через какое-то время (Я дней пять спала часа по четыре, когда влезала на велафакс, 0/10, но потом стал...",1.0,RuDReC,tweet,ru,SMM4H_26_train
46732,ru_12008,"из-за снятия кветиапина побочка от велаксина усилилась и теперь я вообще не могу есть - вкус пищи отвратительный, внешний вид не привлекает и вообще уберите это от меня.",1.0,RuDReC,tweet,ru,SMM4H_26_train
46735,en_4621,"@USER______ effexor , second time was a combination of some other prescribed pills and Pristiq",0.0,previous_SMM4H,tweet,en,SMM4H_26_train


In [95]:
len(train_val)

29305

In [96]:
train_val["language"].value_counts()

language
en    17128
ru    10695
de     1482
Name: count, dtype: int64

In [97]:
train_val["label"].value_counts()

label
0.0    26942
1.0     2363
Name: count, dtype: int64

In [98]:
test = dev_data[dev_data["language"].isin(selected_langs)].copy()
test

,id,text,label,origin,type,language,split
0,de_456,"Hallo <user>, DCIS = duktales Carcinoma in situ. Es gibt noch das LCIS = lobuläres Carcinoma in situ. In situ bedeutet an Ort und Stelle. Das sind Brustkrebsvorstufen. Manche Ärzte sprechen lieber...",0.0,LREC2022,forum post,de,SMM4H_26_dev
4,ru_7420,@pharmazzzevt Акард рабочая штука или фуфломицин? Назначили пить после курса Ксарелто и очистки легких от ТЭЛА.,0.0,RuDReC,tweet,ru,SMM4H_26_dev
5,de_220,Hallo <user> Es freut mich dass es dir in der Tagesklinik gut geht. Fluoxetin hab ich vor Jahren auch mal verschrieben bekommen..hatte bei mir leider keinen gewünschten Effekt. Aber es reagiert ja...,0.0,LREC2022,forum post,de,SMM4H_26_dev
7,ru_4201,В итоге не рекомендую если заболевание не сильное.,0.0,RuDReC,sentence,ru,SMM4H_26_dev
9,ru_11006,когда второй день пропускаешь приём таблеток и вместо венлафаксина пьёшь алкашку,0.0,RuDReC,tweet,ru,SMM4H_26_dev
...,...,...,...,...,...,...,...
8026,ru_7517,"Лазолван отхаркивающее, пульмикорт снимать приступы и спазмы? У нас есть и то и другое, но основной это лазолван",0.0,RuDReC,tweet,ru,SMM4H_26_dev
8027,ru_10784,"Портал — антидепрессант на основе флуоксетина, объясняет Википедия",0.0,RuDReC,tweet,ru,SMM4H_26_dev
8030,en_171,"I need to put my nuvaring in, it's right next to my wine",0.0,previous_SMM4H,tweet,en,SMM4H_26_dev
8031,de_256,"Liebe <user>, vielen lieben Dank! Das tut mir sehr leid. Ich bin zurzeit mit Hormon Pflaster am experimentieren. Gut geht's mir leider nicht. Bin seit 5 Jahren in der Postmenopause. Ein Ende der B...",0.0,LREC2022,forum post,de,SMM4H_26_dev


In [99]:
len(test)

4191

In [100]:
from src.task1.preprocessor import SMM4HConfig, preprocess

test_clean, col_name = preprocess(
    test,
    text_column="text",
    label_column="label",
    config=SMM4HConfig()
)

2026-02-20 21:59:18,607 - INFO - Starting preprocessing on column: 'text'
2026-02-20 21:59:18,613 - INFO - Applying SMM4H string transformations...
2026-02-20 21:59:18,898 - INFO - Checking for duplicates...
2026-02-20 21:59:18,902 - INFO - Fixing label types in 'label'...


           SMM4H Preprocessing Summary           
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━┓
┃ Metric                                ┃ Value ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━┩
│ Initial Rows                          │  4191 │
│ Rows Removed (Empty)                  │     0 │
│ Rows Removed (Duplicate)              │     3 │
│ Final Rows                            │  4188 │
├───────────────────────────────────────┼───────┤
│ Transformed: HTML Tags Removed        │  1023 │
│ Transformed: User Mentions Anonymized │   570 │
│ Transformed: URLs Anonymized          │    67 │
│ Transformed: Hashtags Cleaned         │   180 │
│ Transformed: Emojis Converted/Deduped │    60 │
│ Transformed: HTML Removed             │     1 │
└───────────────────────────────────────┴───────┘

2026-02-20 21:59:18,904 - INFO - Preprocessing complete.


In [101]:
test_clean

,id,text,label,origin,type,language,split,text_before
0,de_456,"Hallo , DCIS = duktales Carcinoma in situ. Es gibt noch das LCIS = lobuläres Carcinoma in situ. In situ bedeutet an Ort und Stelle. Das sind Brustkrebsvorstufen. Manche Ärzte sprechen lieber von n...",0,LREC2022,forum post,de,SMM4H_26_dev,"Hallo <user>, DCIS = duktales Carcinoma in situ. Es gibt noch das LCIS = lobuläres Carcinoma in situ. In situ bedeutet an Ort und Stelle. Das sind Brustkrebsvorstufen. Manche Ärzte sprechen lieber..."
4,ru_7420,[USER] Акард рабочая штука или фуфломицин? Назначили пить после курса Ксарелто и очистки легких от ТЭЛА.,0,RuDReC,tweet,ru,SMM4H_26_dev,@pharmazzzevt Акард рабочая штука или фуфломицин? Назначили пить после курса Ксарелто и очистки легких от ТЭЛА.
5,de_220,Hallo Es freut mich dass es dir in der Tagesklinik gut geht. Fluoxetin hab ich vor Jahren auch mal verschrieben bekommen..hatte bei mir leider keinen gewünschten Effekt. Aber es reagiert ja bekann...,0,LREC2022,forum post,de,SMM4H_26_dev,Hallo <user> Es freut mich dass es dir in der Tagesklinik gut geht. Fluoxetin hab ich vor Jahren auch mal verschrieben bekommen..hatte bei mir leider keinen gewünschten Effekt. Aber es reagiert ja...
7,ru_4201,В итоге не рекомендую если заболевание не сильное.,0,RuDReC,sentence,ru,SMM4H_26_dev,В итоге не рекомендую если заболевание не сильное.
9,ru_11006,когда второй день пропускаешь приём таблеток и вместо венлафаксина пьёшь алкашку,0,RuDReC,tweet,ru,SMM4H_26_dev,когда второй день пропускаешь приём таблеток и вместо венлафаксина пьёшь алкашку
...,...,...,...,...,...,...,...,...
8026,ru_7517,"Лазолван отхаркивающее, пульмикорт снимать приступы и спазмы? У нас есть и то и другое, но основной это лазолван",0,RuDReC,tweet,ru,SMM4H_26_dev,"Лазолван отхаркивающее, пульмикорт снимать приступы и спазмы? У нас есть и то и другое, но основной это лазолван"
8027,ru_10784,"Портал — антидепрессант на основе флуоксетина, объясняет Википедия",0,RuDReC,tweet,ru,SMM4H_26_dev,"Портал — антидепрессант на основе флуоксетина, объясняет Википедия"
8030,en_171,"I need to put my nuvaring in, it's right next to my wine",0,previous_SMM4H,tweet,en,SMM4H_26_dev,"I need to put my nuvaring in, it's right next to my wine"
8031,de_256,"Liebe , vielen lieben Dank! Das tut mir sehr leid. Ich bin zurzeit mit Hormon Pflaster am experimentieren. Gut geht's mir leider nicht. Bin seit 5 Jahren in der Postmenopause. Ein Ende der Beschwe...",0,LREC2022,forum post,de,SMM4H_26_dev,"Liebe <user>, vielen lieben Dank! Das tut mir sehr leid. Ich bin zurzeit mit Hormon Pflaster am experimentieren. Gut geht's mir leider nicht. Bin seit 5 Jahren in der Postmenopause. Ein Ende der B..."


In [104]:
test_clean[test_clean["language"]=="ru"]["type"].value_counts()

type
tweet       1636
sentence     934
review        99
Name: count, dtype: int64

In [105]:
test_clean[test_clean["language"]=="en"]["type"].value_counts()

type
tweet    885
Name: count, dtype: int64

In [75]:
train_val_clean, col_name = preprocess(
    train_val,
    text_column="text",
    label_column="label",
    config=SMM4HConfig()
)

2026-02-20 15:18:08,379 - INFO - Starting preprocessing on column: 'text'
2026-02-20 15:18:08,488 - INFO - Applying SMM4H string transformations...
2026-02-20 15:18:11,002 - INFO - Checking for duplicates...
2026-02-20 15:18:11,044 - INFO - Fixing label types in 'label'...


           SMM4H Preprocessing Summary           
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━┓
┃ Metric                                ┃ Value ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━┩
│ Initial Rows                          │ 29305 │
│ Rows Removed (Empty)                  │     0 │
│ Rows Removed (Duplicate)              │    61 │
│ Final Rows                            │ 29244 │
├───────────────────────────────────────┼───────┤
│ Transformed: Emojis Converted/Deduped │   478 │
│ Transformed: User Mentions Anonymized │  7942 │
│ Transformed: Hashtags Cleaned         │  3287 │
│ Transformed: HTML Tags Removed        │  2281 │
│ Transformed: URLs Anonymized          │   802 │
└───────────────────────────────────────┴───────┘

2026-02-20 15:18:11,073 - INFO - Preprocessing complete.


In [76]:
train_val_clean

,id,text,label,origin,type,language,split,text_before
0,ru_8872,"знаете эти надписи на баночках с лекарствами звучащие как названия фильмов с подзаголовками? типа «ЦИПРОМЕД ципрофлоксацин 0.3%» вы спросите, что это такое я хз. наверное, киберпанк",0,RuDReC,tweet,ru,SMM4H_26_train,"знаете эти надписи на баночках с лекарствами звучащие как названия фильмов с подзаголовками? типа «ЦИПРОМЕД ципрофлоксацин 0.3%» вы спросите, что это такое я хз. наверное, киберпанк"
1,ru_5231,"Знаю лично людей, которые лечились этим препаратом и успешно избавились от гепатита с. Чтобы лечение было эффективным к этому лекарству еще назначают таблетки рибавирин.",0,RuDReC,sentence,ru,SMM4H_26_train,"Знаю лично людей, которые лечились этим препаратом и успешно избавились от гепатита с. Чтобы лечение было эффективным к этому лекарству еще назначают таблетки рибавирин."
4,en_14557,the following tweets are brought to you by seroquel :trade_mark:,0,previous_SMM4H,tweet,en,SMM4H_26_train,the following tweets are brought to you by seroquel™
5,ru_34,"Я если честно боюсь ребенку столько таблеток в день давать, поэтому принимали обычно три раза в день всего пять дней.",0,RuDReC,sentence,ru,SMM4H_26_train,"Я если честно боюсь ребенку столько таблеток в день давать, поэтому принимали обычно три раза в день всего пять дней."
6,en_11127,"rt [USER]: my breakfast consists of vyvanse, coffee, and more vyvanse.",0,previous_SMM4H,tweet,en,SMM4H_26_train,"rt @USER_______: my breakfast consists of vyvanse, coffee, and more vyvanse."
...,...,...,...,...,...,...,...,...
46727,en_8967,[USER] how long hsve you been on humira,0,previous_SMM4H,tweet,en,SMM4H_26_train,@USER_______ how long hsve you been on humira
46729,ru_11694,"Возможно, у тебя просто побочки из-за того, что ты только садишься на эти колеса. Это пройдет через какое-то время (Я дней пять спала часа по четыре, когда влезала на велафакс, 0/10, но потом стал...",1,RuDReC,tweet,ru,SMM4H_26_train,"Возможно, у тебя просто побочки из-за того, что ты только садишься на эти колеса. Это пройдет через какое-то время (Я дней пять спала часа по четыре, когда влезала на велафакс, 0/10, но потом стал..."
46732,ru_12008,"из-за снятия кветиапина побочка от велаксина усилилась и теперь я вообще не могу есть - вкус пищи отвратительный, внешний вид не привлекает и вообще уберите это от меня.",1,RuDReC,tweet,ru,SMM4H_26_train,"из-за снятия кветиапина побочка от велаксина усилилась и теперь я вообще не могу есть - вкус пищи отвратительный, внешний вид не привлекает и вообще уберите это от меня."
46735,en_4621,"[USER] effexor , second time was a combination of some other prescribed pills and Pristiq",0,previous_SMM4H,tweet,en,SMM4H_26_train,"@USER______ effexor , second time was a combination of some other prescribed pills and Pristiq"


In [77]:
len(train_val_clean)

29244

# Report


### **Task 1.1: Data Selection and Preprocessing**

#### **1. Language Selection & Rationale**

For this study, we selected three languages to investigate cross-lingual transfer capabilities in medical social media analysis:

* **English (`en`):** Selected as the primary high-resource source language ( raw samples).
* **Russian (`ru`):** Selected as a secondary source language (). Including Russian adds linguistic diversity (Cyrillic script, rich morphology), preventing the model from overfitting to Germanic/Latin patterns.
* **German (`de`):** Selected as the **Zero-Shot Target** language. It represents a low-resource scenario in this dataset ( samples) with a very low number of positive cases (only 83 positives). This makes it an ideal candidate for evaluating how well models trained on En/Ru generalize to unseen languages.

#### **2. Dataset Statistics & Class Imbalance**

A major challenge identified in the dataset is severe **class imbalance**. As shown in Table 1, the ratio of positive samples (Adverse Drug Events) is consistently low across all languages.

| Language | Total Samples (Train/Val) | Positive (Class 1) | Negative (Class 0) | Positive Ratio |
| --- | --- | --- | --- | --- |
| **English (`en`)** | 17,128 | 1,198 | 15,930 | **7.0%** |
| **Russian (`ru`)** | 10,695 | 1,082 | 9,613 | **10.1%** |
| **German (`de`)** | 1,482 | 83 | 1,399 | **5.6%** |
| **TOTAL** | **29,305** | **2,363** | **26,942** | **8.1%** |

*Observation:* The model will likely bias towards the negative class. We addressed this during the train/validation split by performing **Stratified Splitting** (stratified by `Language + Label`). This ensured that the validation set maintained the exact same positive ratios (En: 7.0%, Ru: 10.1%) as the training set, preventing skewed evaluation results.

#### **3. Preprocessing Strategy**

We implemented a custom preprocessing pipeline tailored for SMM4H (Social Media Mining for Health). The impact of each step on the training data is quantified below:

* **User Anonymization (@User  `[USER]`):**
* *Impact:* Affected **7,942** records (~27% of data).
* *Justification:* This was the most frequent noise pattern. Anonymization is crucial to strictly adhere to privacy standards and to prevent the model from "overfitting" to specific active users (learning that User X is sick, rather than learning the symptoms).


* **Hashtag Cleaning (`#flu`  `flu`):**
* *Impact:* Affected **3,287** records.
* *Justification:* Hashtags often contain the most important semantic content (keywords). Removing the `#` symbol allows the tokenizer to treat them as standard words, preserving the medical context.


* **HTML & Artifact Removal:**
* *Impact:* Affected **2,281** records.
* *Justification:* Tags like `<br>` break sentence structure. We replaced them with periods (`.`) to restore sentence boundaries for the Transformer model.


* **Emoji Handling (🤢  `:nauseated_face:`):**
* *Impact:* Affected **478** records.
* *Justification:* Emojis are dense information carriers in social media (e.g., a pill or vomit face). Demojizing them into text ensures the BERT model interprets the *sentiment* rather than treating them as unknown tokens.


* **Capitalization (Case Preservation):**
* *Decision:* We explicitly **kept** original capitalization.
* *Justification:* Medical acronyms (e.g., `AIDS`, `ALL`, `US`) are case-sensitive. Lowercasing them would introduce ambiguity (e.g., "aids" as in hearing aids vs. the disease).



#### **4. Final Split Construction**

After preprocessing, the data was split 80/20 into `train.csv` and `val.csv`.

* **Final Train Set:** 23,395 samples (Contains En, Ru, De).
* **Final Val Set:** 5,849 samples.
* **Note:** For the main Task 1 training, we will filter the `train.csv` to strictly exclude German (`de`), reserving it solely for Zero-Shot evaluation.



---

### **Task 1.2: Exploratory Data Analysis & Linguistic Characterization**

#### **1. Dataset Distribution & Balance**

We analyzed the class distribution across the three target languages (`en`, `ru`, `de`) in the training, validation, and test sets.

* **Class Imbalance:**
The dataset exhibits a severe class imbalance. As shown in the `Label Distribution` table, the positive class (Adverse Drug Events) accounts for only **8.1%** of the training data (1,891 positives vs. 21,504 negatives).
* **Russian (`ru`)** has the highest density of positives (**10.1%**).
* **German (`de`)** has the lowest density (**5.6%**), making it a challenging Zero-Shot target due to the scarcity of positive examples even in the source distribution.


* **Language Imbalance:**
English dominates the training corpus (**58.4%**), followed by Russian (**36.6%**). German is a low-resource minority (**5.1%**). This imbalance suggests that a multilingual model will likely bias its internal representations toward English syntax and vocabulary unless addressed (e.g., via resampling or balanced batching).

#### **2. Source Type Analysis**

The data originates from diverse social media platforms, which impacts the linguistic style:

* **Tweets (80.8%):** Short, informal, often containing hashtags and user mentions.
* **Sentences (12.9%):** Likely extracted from longer documents; these tend to be more grammatically standard.
* **Forum Posts (5.1%):** Longer, narrative-style descriptions of personal experiences.

#### **3. Linguistic Characteristics**

We analyzed token lengths (whitespace-split) to understand input complexity.

* **Sequence Lengths:**
* **English & Russian:** Similar average lengths (~16-19 tokens).
* **German:** Significantly longer (**Avg: 100.1 tokens**, Max: 984).
* *Interpretation:* The German subset likely consists of **Forum Posts** rather than Tweets. This drastic difference in length means the model must learn to handle both short, noisy tweets (En/Ru) and long, narrative paragraphs (De). Zero-shot transfer might struggle not just with language, but with the **shift in discourse structure** (short vs. long context).


* **Length vs. Label:**
Positive samples (ADE) are on average **longer** (31.9 tokens) than negative samples (20.9 tokens). This suggests that users describing a side effect tend to provide more narrative detail ("I took X and then Y happened") compared to simple statements ("I need Xanax").

#### **4. Manual Inspection & Qualitative Analysis**

We randomly sampled entries to identify potential challenges:

* **Ambiguity in Negatives:**
* *Example (En):* "I need some ambien or geodon to knock my ass out." -> Labeled **0**.
* *Analysis:* Mentioning a drug for its *intended* purpose (sedation) is not an ADE, but a model might confuse the sentiment "knock my ass out" with a negative effect.


* **Cultural/Linguistic Nuances:**
* *Example (Ru):* A user describes "gnawing nails" due to anxiety and how "Afobazol" helped. This narrative style is complex and requires the model to understand *relief* of symptoms vs. *causation* of symptoms.


* **Translation Quality (En  De):**
* *Source:* "doing Prolia arms."
* *Translation:* "...bekomme Prolia in die Arme."
* *Artifact:* The translation is literal. "Doing Prolia arms" likely refers to an injection site or a specific sensation, which the literal German translation might miss contextually. However, drug names (`Neurontin`, `Paxil`) were correctly preserved.



#### **5. Conclusion**

The EDA reveals three critical challenges for the Zero-Shot task:

1. **Extreme Imbalance:** The target language (De) has very few positives (5.6%).
2. **Domain Shift:** German data is much longer (forum posts) than the source training data (mostly tweets).
3. **Linguistic Diversity:** The model must bridge the gap between English/Russian morphology and German compound structures while handling the shift from short-text to long-text classification.

In [34]:
import pandas as pd
from datasets import DatasetDict, Dataset
from src.utils import get_project_root
def load_data(cfg: dict):
    """
    Loads data from CSVs specified in config.
    Filters by language if 'lang_filter' is provided.
    """
    ds_cfg = cfg["dataset"]
    project_root = get_project_root()

    datasets = {}
    for split in ["train", "val", "test"]:
        path = project_root / ds_cfg[f"{split}_path"]
        if path.exists():
            df = pd.read_csv(path)

            # Ensure text is string and handle NaNs
            df[ds_cfg["text_column"]] = df[ds_cfg["text_column"]].fillna("").astype(str)

            # Filter Languages (Train only on en and ru)
            if "lang_filter" in ds_cfg and split == "train":
                selected_langs = ds_cfg["lang_filter"]
                print(f"Filtering {split} set to languages: {selected_langs}")
                df = df[df['language'].isin(selected_langs)]

            # Rename columns to standard HF names
            df = df.rename(columns={ds_cfg["text_column"]: "text", ds_cfg["label_column"]: "labels"})

            datasets[split] = Dataset.from_pandas(df)
            print(f"Loaded {split}: {len(df)} rows")
        else:
            print(f"[red]Warning: {split} file not found at {path}[/red]")

    return DatasetDict(datasets)

In [35]:
cfg = {"dataset": {
    "train_path": "data/SMM4H_2026/train.csv",
    "val_path": "data/SMM4H_2026/val.csv",
    "test_path": "data/SMM4H_2026/test.csv",
    "text_column": "text",
    "label_column": "label",
    "lang_filter": ["en", "ru"]
  }}
data = load_data(cfg=cfg)

Filtering train set to languages: ['en', 'ru']
Loaded train: 22209 rows
Loaded val: 5849 rows
Loaded test: 4188 rows


In [36]:
data["train"]

Dataset({
    features: ['id', 'text', 'labels', 'origin', 'type', 'language', 'split', 'text_before', '__index_level_0__'],
    num_rows: 22209
})

In [37]:
import numpy as np

print("train:", np.unique(data["train"]["language"]))
print("val:", np.unique(data["val"]["language"]))
print("test:", np.unique(data["test"]["language"]))

train: ['en' 'ru']
val: ['de' 'en' 'ru']
test: ['de' 'en' 'ru']


In [38]:
print("train:", len(data["train"]))
print("val:", len(data["val"]))
print("test:", len(data["test"]))

train: 22209
val: 5849
test: 4188
